_source/_posts/2022-01-13-node-deploy-to-heroku.md

```diff
@@ -0,0 +1,492 @@
+    res.render("users/index", {
+      loggedIn: true,
+      title: "Express",
+      user: req.userContext.userinfo,
+    });
+  });
+
+  return router;
+}
+
+module.exports.userRoutes = userRoutes;
+```
+
+This file is where the magic from the Okta oidc middleware shines.  In your route definition, you are passing in the `oidc.ensureAuthenticated` middleware.  This function will reroute unauthenticated users to the login route you defined in your `app.js` file.  Because you didn't explicitly set this parameter it will direct the user to `/login` and trigger the authentication flow.  The model is passed to the view here also contains some information about the user that you will display to the user when they land on this page.  
+
+### Write some Client Code
+
+Now you can write some client code.  By default, the express-generator sets up some views using `jade`.  Jade is a templating engine for Node.js.  The language incorporates some conditionals and flow controls to make templating HTML from your model easier.  The syntax is fairly simple to learn and there are many HTML to jade converters out there.
+
+Open the `layout.jade` file that was generated with your project and replace the code with the following.
+
+```jade
+doctype html
+html
+  head
+    title= title
+    link(rel='stylesheet' href='/stylesheets/style.css')
+    link(rel='stylesheet' href='css/bootstrap.min.css')
+    script(language='javascript' src='js/bootstrap.min.js')
+
+  body
+    .container
+      header.d-flex.flex-wrap.justify-content-center.py-3.mb-4.border-bottom
+        h5.d-flex.align-items-center.mb-3.mb-md-0.me-md-auto.text-dark.text-decoration-none 
+          a(href="/") Home
+        ul.nav.nav-pills
+          if loggedIn
+            li.nav-item 
+              a(href="users").nav-link Profile
+          li.nav-item
+            if loggedIn
+              form(action="logout" method="post")
+                button(type="submit").nav-link.active Logout
+            else 
+              a(href="Users").nav-link.active Login
+
+      block content
+```
+
+As you can see the HTML elements are used, the classes are chained with a `.` between them, and attributes are bounded by parentheses.  This simple layout takes the `loggedIn` value from the server and determines if it should display a `login` or `logout` button.  It also conditionally shows a tab for the `Profile` page if the user is logged in. 
+
+You can add the profile page by creating a new folder called `users` in your `views` folder and adding a new file called `index.jade`.  Add the following code to it.
+
+```jade
+extends ../layout
+
+block content
+  h1 Welcome #{user.name}
+  P locale: #{user.locale}
+  p UserName: #{user.preferred_username}
+```
+
+The page here displays the user information that was retrieved from Okta when the user logged in.
+
+Finally, replace the code in `views/index.jade` with the following.
+
+```jade
```

Comment:
Can we truncate this file? Since the users have already created the application at this point, the information in this file is redundant.